File: rails/app/data/hand_helper.rb
# frozen_string_literal: true

class HandHelper
  def can_split(hand)
    hand.cards.length == 2 and hand.cards[0].value == hand.cards[1].value
  end

  def can_double_down(hand)
    hand.cards.length == 2
  end

  def highest_total(hand)
    total = hand.cards.inject(0) do |sum, card|
      val = [card.value, 10].min
      if val == 1 and sum + 11 <= 21
        val = 11
      end
      sum + val
    end

    if total > 21
      total = lowest_total(hand)
    end

    total
  end

  def lowest_total(hand)
    hand.cards.inject(0) {|sum, card| sum + card.value}
  end

  def best_total(hand)
    highest_total(hand) <= 21 ? highest_total(hand) : lowest_total(hand)
  end

  def busted?(hand)
    lowest_total(hand) > 21
  end

  def blackjack?(hand)
    highest_total(hand) == 21 and hand.cards.length == 2
  end

  def done?(hand)
    busted?(hand) or hand.is_staying or blackjack?(hand)
  end
end

File: django/blackjack/templatetags/hand_helper.py
from django import template

from blackjack.state.hand import Hand, HandHelper, Result

register = template.Library()

HAND_HELPER = HandHelper()

@register.filter(name="isBlackjack")
def isBlackjack(hand: Hand) -> bool:
    return HAND_HELPER.isBlackjack(hand)

@register.filter(name="isBusted")
def isBusted(hand: Hand) -> bool:
    return HAND_HELPER.isBusted(hand)

@register.filter(name="lowest")
def lowest(hand: Hand) -> int:
    return HAND_HELPER.lowestTotal(hand)

@register.filter(name="highest")
def highest(hand: Hand) -> int:
    return HAND_HELPER.highestTotal(hand)

@register.filter(name="bestTotal")
def bestTotal(hand: Hand) -> int:
    return HAND_HELPER.bestTotal(hand)

@register.filter(name="canHit")
def canHit(hand: Hand) -> bool:
    return not HAND_HELPER.isDone(hand)

@register.filter(name="canStay")
def canStay(hand: Hand) -> bool:
    return canHit(hand)

def buttonsVisible(hand: Hand):
    return hand.result == Result.Playing.__str__()

@register.filter(name="canSplit")
def canSplit(hand: Hand) -> bool:
    return buttonsVisible(hand) and HAND_HELPER.canSplit(hand)

@register.filter(name="canDoubleDown")
def canDoubleDown(hand: Hand) -> bool:
    return buttonsVisible(hand) and HAND_HELPER.canDoubleDown(hand)